src/cockpit/channels/pcp.py

```diff
@@ -0,0 +1,672 @@
+                    raise ChannelError('internal-error', message=str(exc)) from None
+
+                if len(fetched) > 0:
+                    self.send_updates(fetched)
+
+                break
+
+        return total_fetched
+
+    def parse_fetched_results(self, context: 'pmapi.pmContext', results: Any) -> Any:
+        metrics = list(self.metrics)
+        samples: Dict[str, Optional[float] | list[Optional[float]]] = {}
+
+        samples['timestamp'] = float(results.contents.timestamp)
+        for i in range(results.contents.numpmid):
+            values: Optional[List[Optional[float]] | Optional[float]] = None
+            numval: int = results.contents.get_numval(i)
+            metric_desc = self.metric_descriptions[i]
+            content_type = metric_desc.desc.type
+
+            # negative numval is an error code we ignore
+            if numval < 0:
+                samples[metrics[i].name] = None
+                continue
+
+            # Unsupported types
+            if content_type == c_api.PM_TYPE_AGGREGATE or \
+                content_type == c_api.PM_TYPE_EVENT or \
+                content_type == c_api.PM_TYPE_STRING:
+                samples[metrics[i].name] = None
+                continue
+
+            if not metric_desc.instanced:
+                values = self.build_sample(context, results, metric_desc, i, 0)
+            else:
+                vals: List[Optional[float]] = []
+                for j in range(numval):
+                    vals.append(self.build_sample(context, results, metric_desc, i, j))
+                values = vals
+
+            samples[metrics[i].name] = values
+
+        return samples
+
+    def build_sample(self, context, results, metric_desc: MetricInfo, metric: int, instance: int) -> Optional[float]:
+        pmid = results.contents.get_pmid(metric)
+        logger.debug("build_sample pmid=%d, metric=%d, instance=%d", pmid, metric, instance)
+
+        # PCP throws an error when we try to convert a metric with numval <=
+        # instanceid, the C bridge returns NAN. We return the Python equivalent None.
+        valueset = results.contents.get_vset(metric)
+        if valueset.numval <= instance:
+            return None
+
+        valfmt = results.contents.get_valfmt(metric)
+        value = results.contents.get_vlist(metric, instance)
+        content_type = metric_desc.desc.type
+
+        # Make sure we keep the least 48 significant bits of 64 bit numbers
+        # since "delta" and "rate" derivation works on those, and the whole
+        # 64 don't fit into a double.
+        sample_value = None
+        atom = None
+        if content_type == c_api.PM_TYPE_64:
+            try:
+                atom = context.pmExtractValue(valfmt,
+                                              value,
+                                              c_api.PM_TYPE_64,
+                                              c_api.PM_TYPE_64)
+                sample_value = (atom.ll << 16) >> 16
```

Comment:
This makes a strong assumption that Python uses a 64 bit type to compute this. Python has bigints, so that feels a bit brittle to me. Could you use `atom.ll & ((1 << 48) - 1)` to be more robust?